sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/DataSource.scala

```diff
@@ -446,7 +450,12 @@ case class DataSource(
         dataSource.createRelation(
           sparkSession.sqlContext, mode, caseInsensitiveOptions, Dataset.ofRows(sparkSession, data))
       case format: FileFormat =>
-        sparkSession.sessionState.executePlan(planForWritingFileFormat(format, mode, data)).toRdd
+        val qe = sparkSession.sessionState.executePlan(planForWritingFileFormat(format, mode, data))
+        val insertCommand = qe.executedPlan.collect {
+          case w: WrittenFileCommandExec => w
+        }.head
+        insertCommand.cmd.run(sparkSession, insertCommand.children,
+          metricsCallback.getOrElse(_ => ()))
```

Comment:
Because here we need to invoke the command and pass the metrics-updating callback function. `.execute()` on the plan can't allow us do this. I'd think if there is a better way to do this.